### PR TITLE
fail if the provided user name is not k8s compliant

### PR DESCRIPTION
**Pull request by MatousJobanek**
we should fail if the provided name of the user is not compliant with k8s names and label values.

### CODE REVIEW

**Review by xcoulon — APPROVED** [review]

**Comment by xcoulon** [comment]
> we should fail if the provided name of the user is not compliant with k8s names and label values.

it's much simpler and much better than sanitizing the username, indeed!

**Review by mfrancisc — COMMENTED** [review]
Thanks for implementing this check 🙏 

I have few minor questions/clarifications.

File: pkg/cmd/generate/cluster_test.go [hunk 1]
```diff
@@ -238,3 +259,18 @@ func newKubeSawAdminsWithDefaultClusters(serviceAccounts []assets.ServiceAccount
...
+}
+
+func TestSanitizeUserName(t *testing.T) {
+	require.NoError(t, validateUserName("special-name"))
+	require.NoError(t, validateUserName("special.name"))
+	for _, username := range invalidUserNames {
+		err := validateUserName(username)
+		require.Error(t, err)
+		assert.NotContains(t, err.Error(), "label")
+		assert.NotContains(t, err.Error(), "subdomain")
```
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/cmd/generate/cluster_test.go:274]: should we check that the error is the expected one ?

File: pkg/cmd/generate/cluster.go [hunk 2]
```diff
@@ -67,3 +74,20 @@ func ensureUsers(ctx *clusterContext, objsCache objectsCache) error {
 
 	return nil
 }
+
+func validateUserName(userName string) error {
+	validationErrors := validation.IsDNS1123Subdomain(userName)
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cmd/generate/cluster.go:79]: we execute two validations:
* `IsDNS1123Subdomain` - which checks  if the string can be used as the name of the resource: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
* `IsValidLabelValue` - which checks if the string can be used as a value of a label:

File: pkg/cmd/generate/cluster_test.go [hunk 3]
```diff
@@ -238,3 +259,18 @@ func newKubeSawAdminsWithDefaultClusters(serviceAccounts []assets.ServiceAccount
 		serviceAccounts,
 		users)
 }
+
+var invalidUserNames = []string{
+	"special#-name", "special:-name", "-name", "name-", "special_name", "specialName",
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cmd/generate/cluster_test.go:264]: nope, you can create a resource only with lower-case characters.

File: pkg/cmd/generate/cluster_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cmd/generate/cluster_test.go]: added some comments here 7f3b039

**Review by mfrancisc — APPROVED** [review]
Great job!

Thanks for answering my questions 🙏

File: pkg/cmd/generate/cluster.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/cmd/generate/cluster.go:80]: ok I see, I got confused by looking at the `TransformUsername` function in toolchain common, where we just use the `IsDNS1123Label` . Thanks for explaining this!

File: pkg/cmd/generate/cluster_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/cmd/generate/cluster_test.go:264]: got it! thanks!

File: pkg/cmd/generate/cluster_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/cmd/generate/cluster_test.go]: makes sense! thanks for the additional comments!

File: pkg/cmd/generate/cluster_test.go [hunk 4]
```diff
@@ -230,6 +231,26 @@ func TestUsers(t *testing.T) {
...
+					Users(User("john+user", []string{"12345"}, true, "", permissionsForAllNamespaces...)),
+				)
+				ctx := newAdminManifestsContextWithDefaultFiles(t, kubeSawAdmins)
+				clusterCtx := newFakeClusterContext(ctx, configuration.Host)
+
+				// when
+				err := ensureUsers(clusterCtx, objectsCache{})
+
+				// then
+				require.Error(t, err)
```
Comment (alexeykazakov) [inline pkg/cmd/generate/cluster_test.go:250]: I usually prefer checking the actual error just to make sure the error message is informative and it's actually the expected error and not something else. Unless it's tricky to hardcode the expected error message in the test.

**Review by alexeykazakov — APPROVED** [review]
Thanks for the fix!

File: pkg/cmd/generate/cluster_test.go (hunk 4 above)
Comment (fbm3307) [reply pkg/cmd/generate/cluster_test.go:250]: +1  to checking the actual error

**Review by fbm3307 — APPROVED** [review]
Thanks for this, i liked this approach

File: pkg/cmd/generate/cluster_test.go (hunk 4 above)
Comment (MatousJobanek) [reply pkg/cmd/generate/cluster_test.go:250]: see https://github.com/kubesaw/ksctl/pull/94#discussion_r1888580456

**Pull request merged**